### DISCUSSION TITLE
Using TouchArea with Image { image-fit: contain; } inconvenient. Am I missing something?

### DISCUSSION THREAD

**Discussion by kaarelen** [slint-ui/slint#4473] [General] [answered]
Using TouchArea with Image { image-fit: contain; } inconvenient. Am I missing something? Probably there is a workaround for this case \ or prob it should be made for first time?

In simple words - TouchArea sticks outside Image.

example: TouchArea placed inside 
```
Image {
            image-fit: contain;
            width: parent.width;
            height: parent.height;
}
```
will occupy full size of the Image, so it's tricky to map a mouse event to the same pixel on the image, if any gaps surrounds image (and they will - because of  "image-fit: contain;")

```
Image {
    source: get_image();

    width: parent.width;
    height: parent.height;
    image-fit: contain;

    TouchArea {
        Rectangle {
            background: red;
            opacity: 50%;
        }
    }
}
```
![image](https://github.com/slint-ui/slint/assets/57552528/e456a6c6-e431-4fda-9385-3c9b672c8157)

PS: sorry for my level of English, if something is not clear.

**Comment by kaarelen** [comment]
PPSS: I know that I can map mouse events, because I know size of the window & aspect ratio of the image,
but I will prefer a solution, not a workaround

**Comment by W9pi3cZ1** [comment]
You just need to set the height or width of Image.

***Note*: Just change one of them, the other will automatically be proportional. Otherwise, the height and width will be changed to what you set, even if it doesn't fit the proportions.**

**Comment by kaarelen** [comment]
Hi, @W9pi3cZ1, thank you for response!
I'm expecting behavior like that (look at the screenshots), as I understand it's impossible to achieve it with your solution (just tested it, it doesn't work as expected at all). Is my understanding correct?

![image](https://github.com/slint-ui/slint/assets/57552528/9a92b39c-6776-4c45-a734-9670f86c983e)
![image](https://github.com/slint-ui/slint/assets/57552528/141e503d-60ac-4e64-b888-07882527bdc7)

**Comment by bjorn** [comment]
I think the `image-fit` property is not suitable for what you're trying to achieve. This is because it only affects the position and size at which the image is rendered _within the `Image` item_. If you need the user to interact with the image and conveniently know which location on the image they clicked, you should instead use the `width` and `height` properties to make sure your image has the right size.

To avoid a binding loop, you could probably have single expression determine the scale the image should have, based on its parent width and height, and then set the image's width and height based on the scale. You need to make sure though, that the parent size does not also depend on the image size.

**Reply by kaarelen** [reply]
Oh, haven't noticed your message, I'm sorry.

Could you check out my [last comment](https://github.com/slint-ui/slint/discussions/4473#discussioncomment-8314885)?

If I've understood you correctly - I need to forget about resizing, like on the video, if I want to get simple user-interaction code?

**Comment by W9pi3cZ1** [comment]
I trying to Implement this. It's works.
![截图_2024-01-30_21-17-37](https://github.com/slint-ui/slint/assets/116885083/993bd70b-59a2-4b87-a500-5dc786d4bc22)
![截图_2024-01-30_21-19-01](https://github.com/slint-ui/slint/assets/116885083/9808dd79-e166-4b2b-a150-201b0d0d0f38)
![截图_2024-01-30_21-19-19](https://github.com/slint-ui/slint/assets/116885083/6d139d00-3607-462a-b9dc-58f692fad39d)

**Comment by kaarelen** [comment]
Hello, @W9pi3cZ1. I'm really grateful for the time you have spent, but I've just tried your approach and it doesn't work for me as I've expected - the image doesn't stay fitted to the window size and can overflow (watch the attached video). 
Behavior of your code - 
[recording (1).webm](https://github.com/slint-ui/slint/assets/57552528/cb914aed-bed0-4239-b515-07d5378687a7)


I want to make, as clear as possible, what I want to get. So here is video of expected behavior (in terms of resizing) - 
[recording.webm](https://github.com/slint-ui/slint/assets/57552528/fed8e631-1b2d-4e06-9785-fe8b50ac001d)

Actual problem - I want to fit TouchArea into a RENDERED image without loosing this fancy resizing, but I don't have any access to real rendered(!) image size \ position :
1) "image.source.height" relates to real (file) image size.
2) "image.height" relates to size of the box that contains rendered image.
3) "image.padding" |  "padding-bottom" | etc - stays on 0px.

So there is no clear way to get this information.

Here is a simplified example of my code:
```
export component MainWindow inherits Window{
    pure callback get_image() -> image;

    no-frame: true;    
    image := Image {
        // source: get_image();
        source: @image-url("sample.jpeg");
        width: parent.width;
        height: parent.height;

        image-fit: contain;
        image-rendering: smooth;
        // image-rendering: pixelated;

        touch := TouchArea {
            width: image.width;
            height: image.height;

            Rectangle {
                background: red;
                opacity: 30%;
            }
        }
    }
}
```

Here is a code used for video of "your" code:
```
export component MainWindow inherits Window{
    pure callback get_image() -> image;
    pure callback save_area(
        int,
        int,
        int,
        int,
    );
    
    no-frame: true;

    default-font-size: 24px;
    Image {
        source: @image-url("sample.jpeg");
        // property <float> ratio: self.source.width / (self.width / 1px);
        width: parent.width;
        image-fit: contain;
        touch := TouchArea {
            Rectangle {
                background: red;
                opacity: 30%;
            }
        }
        Rectangle {
            x: touch.pressed-x;
            y: touch.pressed-y;
            background: blue;
            width: 1px;
            height: 1px;
        }
    }
}
```

Code directly from dreams:
```
Image {
    source: @image-url("sample.jpeg");
    image-fit: contain;
    touch := TouchArea {
        width: parent.inner.width;
    }
}
```

PS: I can just calculate rendered size of the image from window size, image source size, aspect ratio of the image, but It's definitely NOT a convenient way.

**Reply by ogoffart** [reply]
I think for your usecase it is better not to use `image-fit: contain;` but instead compute the size of your image manually using the ratio.

Something like this (untested)

```
Rectangle {
   Image {
      source: ...
      property <float> ratio: self.source.width / self.source.height;
      width: min(parent.width, parent.height * ratio);
      height: min(parent.height, parent.width / ratio);
      TouchArea { ... }
   }
}
```

**Reply by kaarelen** [reply]
@ogoffart works as charm! Thanks & have a good day!

**Reply by kaarelen** [reply]
@ogoffart I've rushed, while I had to take my time.
So, this approach is VERY laggy on bigger images, while "image-fit: contain" works smooth.
It is still a solution. But not in my case, I need responsive interface.

**Reply by ogoffart** [reply]
I can't think of a reason why the performance should be any different.

**Reply by kaarelen** [reply]
@ogoffart yeah, agreed, but reality hits)

Here are minimal examples to reproduce

Laggy, barely usable:
```
export component MainWindow inherits Window{
    //pure callback get_image() -> image;
    image := Image {
        source: @image-url("test.png");
        property <int> ratio: image.source.width / image.source.height;
        width: min(parent.width, parent.height * ratio);
        height: min(parent.height, parent.width / ratio);

        touch := TouchArea {
            Rectangle {
                background: red;
                opacity: 30%;
            }
        }
    }
}
```

Fast and responsive:
```
export component MainWindow inherits Window{
    //pure callback get_image() -> image;
    image := Image {
        source: @image-url("test.png");
        image-fit: contain;
        width: parent.width;
        height: parent.height;

        touch := TouchArea {
            Rectangle {
                background: red;
                opacity: 30%;
            }
        }
    } 
}
```
[test.png](https://github.com/slint-ui/slint/assets/57552528/da265201-7aab-4f36-919b-796f0e799544) if you need sample image
size of the image - 2560x1440

**Reply by ogoffart** [reply]
For me the two are the same.  
What renderer are you using?  Is it a release or a debug build?
You can use set an env variable `SLINT_DEBUG_PERFORMANCE=console,refresh_full_speed,overlay` to debug performance.

**Reply by kaarelen** [reply]
It's still laggy.

```
> SLINT_DEBUG_PERFORMANCE=console,refresh_full_speed,overlay cargo run --release

Slint: Build config: release; Backend: Qt backend (platform wayland)
average frames per second: 53 details from last frame: [0 layers created]
average frames per second: 63 details from last frame: [0 layers created]
average frames per second: 19 details from last frame: [0 layers created]
average frames per second: 24 details from last frame: [0 layers created]
average frames per second: 19 details from last frame: [0 layers created]
average frames per second: 19 details from last frame: [0 layers created]
average frames per second: 20 details from last frame: [0 layers created]
average frames per second: 25 details from last frame: [0 layers created]
```

**Reply by kaarelen** [reply, marked as answer]
@ogoffart final solution - forget about image and just scale size of TouchArea same way, as u suggested before for image. In my case its whole overlay, not just TouchArea, but idea is the same.

Thoughts about why u had not any issues with that, and I had - probably u have a lot more powerful device, so in your case performance is degrading too, but fps still higher than 60 and my 10 years old pc with weak CPU just dies from this load. So that's why.

Writing here just to provide info for next generations (for me, when i will forget this solution 0)0))

**Reply by kaarelen** [reply]
here is minimal code that's enough to understand solution in few seconds

```
image := Image {
        source: get_image();
        // source: @image-url("test.png");

        width: parent.width;
        height: parent.height;
        image-fit: contain;

        image_overlay := Rectangle {
              property <int> ratio: image.source.width / image.source.height;
              width: min(image.width, image.height * ratio);
              height: min(image.height, image.width / ratio);
        }
}
```

**Reply by bjorn** [reply]
That's clever, thanks for posting your solution!
